sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala

```diff
@@ -1047,6 +1047,57 @@ case class ToDate(child: Expression) extends UnaryExpression with ImplicitCastIn
 }
 
 /**
+ * Parses a column with a format to a timestamp.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(timestamp_str, fmt) - Parses the `timestamp` expression " +
+    "with the `fmt` expression.",
+  extended = """
+    Examples:
+      > SELECT _FUNC_('2016-12-31', 'yyyy-MM-dd');
+       2016-12-31
+             """)
+case class ParseToDate(left: Expression, right: Expression, child: Expression)
+  extends RuntimeReplaceable with ImplicitCastInputTypes {
+
+  def this(left: Expression, format: Expression) = {
+    this(left, format, Cast(new ParseToTimestamp(left, format), DateType))
+  }
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(StringType, StringType)
+  override def flatArguments: Iterator[Any] = Iterator(left, right)
+  override def sql: String = s"$prettyName(${left.sql}, ${right.sql})"
+
+  override def prettyName: String = "to_date"
+  override def dataType: DataType = DateType
+}
+
+/**
+ * Parses a column with a format to a timestamp.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(timestamp, fmt) - Parses the `timestamp` expression with the `fmt` expression.",
+  extended = """
+    Examples:
+      > SELECT _FUNC_('2016-12-31', 'yyyy-MM-dd');
+       2016-12-31 00:00:00.0
+  """)
+case class ParseToTimestamp(left: Expression, right: Expression, child: Expression)
+  extends RuntimeReplaceable with ImplicitCastInputTypes {
+
+  def this(left: Expression, format: Expression) = {
+  this(left, format, Cast(UnixTimestamp(left, format), TimestampType))
+}
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(StringType, StringType)
+  override def flatArguments: Iterator[Any] = Iterator(left, right)
+  override def sql: String = s"$prettyName(${left.sql}, ${right.sql})"
+
+  override def prettyName: String = "to_timestamp"
+  override def dataType: DataType = TimestampType
```

Comment:
If I remove this `DataType` override I will end up with the right answer but the wrong type (string).

If I have this `DataType` override I will end up with the wrong answer but the right type.

If I cast to the expected type by either having the `dataType` override or removing it, I will get the expected output. But I have to cast after performing `to_timestamp` or `to_date`. The cast within the expression does not seem to be respected.